libvast/src/meta_index.cpp

```diff
@@ -13,67 +13,194 @@
+      VAST_DEBUG(this, "could not create a synopsis for layout:", layout);
+      blacklisted_layouts_.insert(layout);
+    }
+  }
+  VAST_ASSERT(table_syn->size() == slice.columns());
+  for (size_t col = 0; col < slice.columns(); ++col) {
+    if (auto& syn = (*table_syn)[col])
+      for (size_t row = 0; row < slice.rows(); ++row)
+        syn->add(*slice.at(row, col));
+  }
 }
 
-// -- properties ---------------------------------------------------------------
+std::vector<uuid> meta_index::lookup(const expression& expr) const {
+  VAST_ASSERT(!caf::holds_alternative<caf::none_t>(expr));
+  // TODO: we could consider a flat_set<uuid> here, which would then have
+  // overloads for inplace intersection/union and simplify the implementation
+  // of this function a bit.
+  using result_type = std::vector<uuid>;
+  auto all_partitions = [&] {
+    result_type result;
+    result.reserve(partition_synopses_.size());
+    std::transform(partition_synopses_.begin(),
+                   partition_synopses_.end(),
+                   std::back_inserter(result),
+                   [](auto& x) { return x.first; });
+    std::sort(result.begin(), result.end());
+    return result;
+  };
+  return caf::visit(detail::overload(
+    [&](const conjunction& x) -> result_type {
+      VAST_ASSERT(!x.empty());
+      auto i = x.begin();
+      auto result = lookup(*i);
+      if (!result.empty())
+        for (++i; i != x.end(); ++i) {
+          auto xs = lookup(*i);
+          if (xs.empty())
+            return xs; // short-circuit
+          detail::inplace_intersect(result, xs);
+        }
+      return result;
+    },
+    [&](const disjunction& x) -> result_type {
+      result_type result;
+      for (auto& op : x) {
+        auto xs = lookup(op);
+        if (xs.size() == partition_synopses_.size())
+          return xs; // short-circuit
+        detail::inplace_unify(result, xs);
+      }
+      return result;
+    },
+    [&](const negation&) -> result_type {
+      // We cannot handle negations, because a synopsis may return false
+      // positives, and negating such a result may cause false negatives.
+      return all_partitions();
+    },
+    [&](const predicate& x) -> result_type {
+      // Performs a lookup on all *matching* synopses with operator and data
+      // from the predicate of the expression. The match function uses a record
+      // field to determine whether the synopsis should be queried.
+      auto search = [&](auto match) {
+        VAST_ASSERT(caf::holds_alternative<data>(x.rhs));
+        auto& rhs = caf::get<data>(x.rhs);
+        result_type result;
+        auto found_matching_synopsis = false;
+        for (auto& [part_id, part_syn] : partition_synopses_)
+          for (auto& [layout, table_syn] : part_syn)
+            for (size_t i = 0; i < table_syn.size(); ++i)
+              if (table_syn[i] && match(layout.fields[i])) {
+                found_matching_synopsis = true;
```

Comment:
I think `found_matching_synopsis` can be renamed to `found_matching_fields`. Consequently, the value has to be set to true even if there is no synopsis for that field available. If we take factor out the `synopsis` logic, we can say "a partition is part of the result set if it contains a matching layout". The `synopses` on top is just a means to narrow down that result set further.
It follows that we can `return result;` unconditionally.